desci-server/src/controllers/nodes/publish.ts

```diff
@@ -98,49 +102,270 @@ export const publish = async (req: PublishRequest, res: Response<PublishResBody>
       return res.status(400).json({ error: 'failed' });
     }
 
+    // Check if there is already an ongoing publish job in the queue
     const task = await prisma.publishTaskQueue.findFirst({
       where: { uuid: ensureUuidEndsWithDot(uuid), status: { not: PublishTaskQueueStatus.FAILED } },
     });
 
     if (task) return res.status(400).json({ error: 'Node publishing in progress' });
 
+    let publishTask: PublishTaskQueue | undefined;
+
+    if (useNewPublish) {
+      await syncPublish(
+        ceramicStream,
+        commitId,
+        node,
+        owner,
+        cid,
+        uuid,
+        manifest,
+      );
+    } else {
+      publishTask = await prisma.publishTaskQueue.create({
+        data: {
+          cid,
+          dpid: manifest.dpid?.id,
+          userId: owner.id,
+          transactionId,
+          ceramicStream,
+          commitId,
+          uuid: ensureUuidEndsWithDot(uuid),
+          status: PublishTaskQueueStatus.WAITING,
+        },
+      });
+    };
+
     saveInteraction(
       req,
       ActionType.PUBLISH_NODE,
       {
         cid,
-        dpid: manifest.dpid.id,
+        dpid: manifest.dpid?.id,
         userId: owner.id,
         transactionId,
-        ceramicStream: ceramicStream ?? '',
-        commitId: commitId ?? '',
+        ceramicStream,
+        commitId,
         uuid: ensureUuidEndsWithDot(uuid),
         status: PublishTaskQueueStatus.WAITING,
       },
       owner.id,
     );
 
-    const publishTask = await prisma.publishTaskQueue.create({
-      data: {
-        cid,
-        dpid: manifest.dpid.id,
-        userId: owner.id,
-        transactionId,
-        ceramicStream: ceramicStream ?? '',
-        commitId: commitId ?? '',
-        uuid: ensureUuidEndsWithDot(uuid),
-        status: PublishTaskQueueStatus.WAITING,
-      },
-    });
-
     return res.send({
       ok: true,
-      taskId: publishTask.id,
+      taskId: publishTask?.id,
     });
   } catch (err) {
     logger.error({ err }, '[publish::publish] node-publish-err');
     return res.status(400).send({ ok: false, error: err.message });
-  }
+  };
+};
+
+/**
+ * Synchronously perform publish steps before returning.
+ *
+ * This is generally fast because the ceramic update is already done,
+ * but two cases are a bit slower:
+ * 1. first time publish (wait on backend tx)
+ * 2. dpid upgrade (wait on backend tx)
+ *
+ * Semantically, these can both be made fire-and-forget promises if we can
+ * manage without instantly having the dPID alias available in this function.
+*/
+const syncPublish = async (
+  ceramicStream: string,
+  commitId: string,
+  node: Node,
+  owner: User,
+  cid: string,
+  uuid: string,
+  manifest: ResearchObjectV1,
+): Promise<void> => {
+  const logger = parentLogger.child({
+    module: 'NODE::syncPublish',
+    uuid,
+    cid,
+    ceramicStream,
+    commitId,
+  });
+
+  const latestNodeVersion = await prisma.nodeVersion.findFirst({
```

Comment:
just a note that nodeVersion in DB doesn't just include published changes but all draft changes

Reply:
This is just moved around, haven't written this. But I do think this lot smells a bit fishy, can't entirely follow the logic and implications:

```ts
  const latestNodeVersion = await prisma.nodeVersion.findFirst({
    where: {
      id: -1,
      nodeId: node.id,
    },
    orderBy: {
      id: "desc",
    },
  });

  // Prevent duplicating the NodeVersion entry if the latest version is the same as the one we're trying to publish, as a draft save is triggered before publishing
  const latestNodeVersionId = latestNodeVersion?.manifestUrl === cid
    ? latestNodeVersion.id
    : -1;

  const nodeVersion = await prisma.nodeVersion.upsert({
    where: {
      id: latestNodeVersionId,
    },
    update: {
      commitId,
    },
    create: {
      nodeId: node.id,
      manifestUrl: cid,
      commitId,
    },
  });
```